adapters/tlsio_appleios.c

```diff
@@ -0,0 +1,546 @@
+#include <ctype.h>
+#include <stdio.h>
+#include <string.h>
+#include "azure_c_shared_utility/gballoc.h"
+#include "azure_c_shared_utility/tlsio_appleios.h"
+#include "azure_c_shared_utility/socketio.h"
+#include "azure_c_shared_utility/optimize_size.h"
+#include "azure_c_shared_utility/xlogging.h"
+#include "azure_c_shared_utility/platform.h"
+#include "azure_c_shared_utility/threadapi.h"
+#include "azure_c_shared_utility/optionhandler.h"
+
+#include <CoreFoundation/CoreFoundation.h>
+#include <CFNetwork/CFSocketStream.h>
+
+/* Codes_SRS_TLSIO_APPLEIOS_32_001: [ The tlsio_appleios shall implement and export all the Concrete functions in the VTable IO_INTERFACE_DESCRIPTION defined in the `xio.h`. ]*/
+/* Codes_SRS_TLSIO_APPLEIOS_32_002: [ The tlsio_appleios shall report the open operation status using the IO_OPEN_RESULT enumerator defined in the `xio.h`.]*/
+/* Codes_SRS_TLSIO_APPLEIOS_32_003: [ The tlsio_appleios shall report the send operation status using the IO_SEND_RESULT enumerator defined in the `xio.h`. ]*/
+/* Codes_SRS_TLSIO_APPLEIOS_32_004: [ The tlsio_appleios shall call the callbacks functions defined in the `xio.h`. ]*/
+#include "azure_c_shared_utility/xio.h"
+
+/* Codes_SRS_TLSIO_APPLEIOS_32_005: [ The tlsio_appleios shall received the connection information using the TLSIO_CONFIG structure defined in `tlsio.h`. ]*/
+#include "azure_c_shared_utility/tlsio.h"
+
+
+/* Codes_SRS_TLSIO_APPLEIOS_32_001: [ The tlsio_appleios shall implement and export all the Concrete functions in the VTable IO_INTERFACE_DESCRIPTION defined in the `xio.h`. ]*/
+CONCRETE_IO_HANDLE tlsio_appleios_create(void* io_create_parameters);
+void tlsio_appleios_destroy(CONCRETE_IO_HANDLE tls_io);
+int tlsio_appleios_open(CONCRETE_IO_HANDLE tls_io, ON_IO_OPEN_COMPLETE on_io_open_complete, void* on_io_open_complete_context, ON_BYTES_RECEIVED on_bytes_received, void* on_bytes_received_context, ON_IO_ERROR on_io_error, void* on_io_error_context);
+int tlsio_appleios_close(CONCRETE_IO_HANDLE tls_io, ON_IO_CLOSE_COMPLETE on_io_close_complete, void* callback_context);
+int tlsio_appleios_send(CONCRETE_IO_HANDLE tls_io, const void* buffer, size_t size, ON_SEND_COMPLETE on_send_complete, void* callback_context);
+void tlsio_appleios_dowork(CONCRETE_IO_HANDLE tls_io);
+int tlsio_appleios_setoption(CONCRETE_IO_HANDLE tls_io, const char* optionName, const void* value);
+OPTIONHANDLER_HANDLE tlsio_appleios_retrieveoptions(CONCRETE_IO_HANDLE tls_io);
+
+static const IO_INTERFACE_DESCRIPTION tlsio_handle_interface_description =
+{
+    tlsio_appleios_retrieveoptions,
+    tlsio_appleios_create,
+    tlsio_appleios_destroy,
+    tlsio_appleios_open,
+    tlsio_appleios_close,
+    tlsio_appleios_send,
+    tlsio_appleios_dowork,
+    tlsio_appleios_setoption
+};
+
+/* Codes_SRS_TLSIO_APPLEIOS_32_027: [ The tlsio_appleios_open shall set the tlsio to try to open the connection for 10 times before assuming that connection failed. ]*/
+#define MAX_TLS_OPENING_RETRY  10
```

Comment:
>MAX_TLS_OPENING_RETRY [](start = 8, length = 21)

Why is a retry used at this layer?

Reply:
I don't think it should be here at all.

---
In reply to: [139022540](https://github.com/Azure/azure-c-shared-utility/pull/109#discussion_r139022540) [](ancestors = 139022540)